debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/AbstractTransactionCachingLogMinerEventProcessor.java

```diff
@@ -0,0 +1,430 @@
+     * This is because the method needed {@code entrySet()} is made available on two different concrete
+     * interfaces between the embedded and remote cache implementations, and therefore we need to access
+     * this method from the concrete implementation classes (RemoteCache and Cache) rather than from
+     * the common class used by CacheProvider (BasicCache).
+     *
+     * @param minCacheScn the minimum system change number to keep entries until
+     */
+    protected void purgeCache(Scn minCacheScn) {
+        getProcessedTransactionsCache().removeIf(entry -> Scn.valueOf(entry.getValue()).compareTo(minCacheScn) < 0);
+        getSchemaChangesCache().removeIf(entry -> Scn.valueOf(entry.getKey()).compareTo(minCacheScn) < 0);
+    }
+
+    @Override
+    protected Scn calculateNewStartScn(Scn endScn, Scn maxCommittedScn) throws InterruptedException {
+
+        // Cleanup caches based on current state of the transaction cache
+        final Optional<T> oldestTransaction = getOldestTransactionInCache();
+        final Scn minCacheScn;
+        final Instant minCacheScnChangeTime;
+        if (oldestTransaction.isPresent()) {
+            minCacheScn = oldestTransaction.get().getStartScn();
+            minCacheScnChangeTime = oldestTransaction.get().getChangeTime();
+        }
+        else {
+            minCacheScn = Scn.NULL;
+            minCacheScnChangeTime = null;
+        }
+
+        if (!minCacheScn.isNull()) {
+            abandonTransactions(getConfig().getLogMiningTransactionRetention());
+            purgeCache(minCacheScn);
+        }
+        else {
+            getSchemaChangesCache().removeIf(e -> true);
+        }
+
+        if (getConfig().isLobEnabled()) {
+            if (getTransactionCache().isEmpty() && !maxCommittedScn.isNull()) {
+                offsetContext.setScn(maxCommittedScn);
+                dispatcher.dispatchHeartbeatEvent(partition, offsetContext);
+            }
+            else {
+                if (!minCacheScn.isNull()) {
+                    getProcessedTransactionsCache().removeIf(entry -> Scn.valueOf(entry.getValue()).compareTo(minCacheScn) < 0);
+                    offsetContext.setScn(minCacheScn.subtract(Scn.valueOf(1)));
+                    dispatcher.dispatchHeartbeatEvent(partition, offsetContext);
+                }
+            }
+            return offsetContext.getScn();
+        }
+        else {
+
+            if (!getLastProcessedScn().isNull() && getLastProcessedScn().compareTo(endScn) < 0) {
+                // If the last processed SCN is before the endScn we need to use the last processed SCN as the
+                // next starting point as the LGWR buffer didn't flush all entries from memory to disk yet.
+                endScn = getLastProcessedScn();
+            }
+
+            // update offsets
+            offsetContext.setScn(endScn);
+            metrics.setOldestScnDetails(minCacheScn, minCacheScnChangeTime);
+            metrics.setOffsetScn(endScn);
```

Comment:
```suggestion
            offsetContext.setScn(minCacheScn.isNull() ? endScn : minCacheScn.subtract(Scn.valueOf(1)));
            metrics.setOldestScnDetails(minCacheScn, minCacheScnChangeTime);
            metrics.setOffsetScn(offsetContext.getScn());
```

Reply:
I need to test this with Infinispan, but I'm pretty confident this should likely solve both problems.  I'm actually shocked, however, this wasn't identified previously :thinking: